[WIP] Wechat oauth (first draft)

lhr0909 opened the pull request:
<!-- INSTRUCTION: Your Pull Request name should start with one of the following tags -->
<!-- [NEW] For new features -->
<!-- [FIX] For bug fixes -->
<!-- [BREAK] For pull requests including breaking changes -->

<!-- INSTRUCTION: Inform the issue number that this PR closes, or remove the line below -->
attempting to close https://github.com/RocketChat/Rocket.Chat/issues/7576

<!-- INSTRUCTION: Link to a https://github.com/RocketChat/docs PR with added/updated documentation or an update to the missing/outdated documentation list, see https://rocket.chat/docs/contributing/documentation/  -->

<!-- INSTRUCTION: Tell us more about your PR with screen shots if you can -->

We followed [Chinese version of this documentation](https://open.weixin.qq.com/cgi-bin/showdocument?action=dir_list&t=resource/res_list&verify=1&id=open1419316505&token=&lang=en_US) to integrate WeChat OAuth. However, we weren't too familiar with the entire integration process as of yet, so we only managed to hack the custom OAuth to support WeChat OAuth process.

Basically, WeChat's OAuth is tied to an Official Account. There is an extra step to retrieve a `code` from the WeChat server, in order to verify that the official account has the permissions to use OAuth, and we need to use the said `code` in order to request `access_token`. and get user info.

Here is the custom OAuth ingress.

![image](https://user-images.githubusercontent.com/1154575/44194003-4b30e880-a167-11e8-91f0-82fea230269b.png)

QR code verification that comes from WeChat, needs to scan with WeChat App to authorize: 
![image](https://user-images.githubusercontent.com/1154575/44194037-5e43b880-a167-11e8-9843-b9f51503bd68.png)

Basically we would like the RocketChat team to give us pointers and see how to properly integrate into RocketChat as a supported OAuth method.

engelgabriel committed:
Updated icon assets

tassoevan committed:
[FIX] Jump to message search result action (#11613)

rodrigok committed:
[FIX] HipChat importer wasn’t compatible with latest exports (#11597)

sampaiodiego committed:
Bump version to 0.68.1

sampaiodiego committed:
Merge pull request #11616 from RocketChat/release-0.68.1

Release 0.68.1

c0dzilla committed:
[FIX] Incorrect migration version in v130.js (#11544)

sampaiodiego committed:
Bump version to 0.68.2

sampaiodiego committed:
Merge pull request #11630 from RocketChat/release-0.68.2

Release 0.68.2

Hudell committed:
[FIX] Missing chat history for users without permission `preview-c-room` (#11639)

Commit:
[FIX] User info APIs not returning customFields correctly (#11625)

rndmh3ro committed:
[FIX] Prune translations in German (#11631)

sampaiodiego committed:
[FIX] Prune translation on room info panel (#11635)

Hudell committed:
[FIX] SAML login not working when user has multiple emails (#11642)

sampaiodiego committed:
Bump version to 0.68.3

sampaiodiego committed:
Merge pull request #11650 from RocketChat/release-0.68.3

Release 0.68.3

vynmera committed:
[FIX] Regression in prune by user, and update lastMessage (#11646)

* Fix regression

* Also update lastMessage

* code style fix

sampaiodiego committed:
Fix setup wizard logo

tassoevan committed:
Merge pull request #11708 from RocketChat/fix-setup-wizard-logo

[FIX] Broken logo on setup wizard

tassoevan committed:
Merge pull request #11708 from RocketChat/fix-setup-wizard-logo

[FIX] Broken logo on setup wizard

sampaiodiego committed:
[FIX] Closed connections being storing on db (#11709)

sampaiodiego committed:
[FIX] Default server language not being applied (#11719)

* Fix default server language not being applied

* Always set user's language on localStorage

sampaiodiego committed:
[FIX] Default server language not being applied (#11719)

* Fix default server language not being applied

* Always set user's language on localStorage

sampaiodiego committed:
Bump version to 0.68.4

sampaiodiego committed:
Merge pull request #11743 from RocketChat/release-0.68.4

Release 0.68.4

c19 committed:
wip

c19 committed:
wechat oauth ok

crazy-max committed:
[IMPROVE] Add template tag #{userdn} to filter LDAP group member format (#11662)

ggazzo committed:
[FIX] message attachments (padding, order and short) (#11740)

[FIX] Message attachments was not respecting sort and lost spacing

ggazzo committed:
[WIP] Messagebox fix performance (#11686)

<!-- INSTRUCTION: Your Pull Request name should start with one of the following tags -->
<!-- [NEW] For new features -->
<!-- [FIX] For bug fixes -->
<!-- [BREAK] For pull requests including breaking changes -->

<!-- INSTRUCTION: Inform the issue number that this PR closes, or remove the line below -->
should close #10583

<!-- INSTRUCTION: Link to a https://github.com/RocketChat/docs PR with added/updated documentation or an update to the missing/outdated documentation list, see https://rocket.chat/docs/contributing/documentation/  -->

<!-- INSTRUCTION: Tell us more about your PR with screen shots if you can -->

glstewart17 committed:
[FIX] Render Attachment Pretext When Markdown Specified (#11578)

Render pretext as markdown when mrkdown_in contains "pretext"
Closes #3828

![image](https://user-images.githubusercontent.com/41751617/43274277-c5de8e30-90cc-11e8-8ba5-a57ad4af2571.png)

rssilva committed:
[FIX] Fixing timeAgo function on directory (#11728)

Fix #11663 

Adding `timeAgo` function to a `helpers` file so we can test it because just importing it from the previous file will cause an error `Template is not defined`. I reproduced the mentioned case on a test.

vynmera committed:
[FIX] Sound upload not working in IE and FF (#11139)

vynmera committed:
Fix code duplicating characters (#11467)

Atisom committed:
minor modification (#11761)

tassoevan committed:
[FIX] Escape meta data before inject in head tag (#11730)

gsperezb committed:
[FIX] wrong create date of channels and users on directory view (#11682)

#11663 
Compares day, month and year to show the correct create date of channels and users.

renatobecker committed:
[FIX] "User is typing" not working in new Livechat session (#11670)

c19 committed:
remove log

lhr0909 committed:
adding dockerfile to build and package using local source

timkinnane committed:
[FIX] Return room ID for groups where user joined (#11703)

timkinnane committed:
[IMPROVE] Role tag UI (#11674)

Role and bot tags are too prominent. When conversations involve lots of sequential messages from bots or users with role tags, the tags make it hard on the eye to track down the message stream. The end effect is a messy view and high cognitive load.

I removed the bot tag from sequential messages entirely and reduced the visual impact of all role tags.

Includes:
- Visual enhancement for less busy tags
- Consolidated styles for role tags
- Use color classes in markup instead of css

Example shows a conversation I'm having with our new buddy Faldo...

Before:
<img width="561" alt="screenshot 2018-08-04 14 52 59" src="https://user-images.githubusercontent.com/1774379/43673494-4ebb4392-9807-11e8-9bc1-ddfc8965569b.png">

After:
<img width="569" alt="screenshot 2018-08-04 16 44 41" src="https://user-images.githubusercontent.com/1774379/43673499-635d75b8-9807-11e8-877d-4a55c723f2bb.png">

engelgabriel committed:
[FIX] Missing twitter:image and og:image tags (#11687)

lhr0909 commented:
This is the OAuth configuration. I think ideally we need to create a panel specifically for WeChat so we can put in the `appid` and `appsecret` information in, and lock in the wechat OAuth URL.

![image](https://user-images.githubusercontent.com/1154575/44195279-28a0ce80-a16b-11e8-9dda-a7fbdaeb79b2.png)

tassoevan committed:
[IMPROVE] Warn about push settings that need server restart (#11784)

tassoevan committed:
[IMPROVE] Escape parameters before send them to email template (#11644)

ggazzo committed:
[FIX] Cannot set property 'input' of undefined (#11775)

Commit:
[NEW] Make font of unread items bolder for better contrast (#8602)

* Make font of unread items bolder for better conrttrast - fixes #8599

* Update sidebar-item.css

sampaiodiego committed:
Add new eslint rules (#11800)

ggazzo committed:
Regression: more tests to improve messagebox autogrow(#11805)

tassoevan committed:
[FIX] Apply Cordova fix in lazy-loaded images sources (#11807)

sampaiodiego committed:
Additional eslint rules  (#11804)

geekgonecrazy committed:
[NEW] Setting to set a JS/CSS CDN (#11779)

Commit:
[FIX] REST endpoints to update user not respecting some settings (#11474)

rodrigok committed:
Run eslint and unit tests on pre-push hook (#11815)

rodrigok committed:
Do not remove package-lock.json of livechat package (#11816)

sampaiodiego committed:
Regression: Fix livechat code issues after new lint rules (#11814)

tassoevan committed:
[FIX] directory search table not clickable lines (#11809)

Hudell committed:
[FIX] SAML is flooding logfile (#11643)

Hudell committed:
[NEW] Setting to block unauthenticated access to avatars (#9749)

Hudell committed:
[FIX] Removed hardcoded values (#11627)

graywolf336 committed:
[FIX] App's i18nAlert is only being displayed as "i18nAlert" (#11802)

* Fix the i18nAlert not being displayed from Apps

* Fix the apps http being invalid when being used as promises

ubarsaiyan committed:
[NEW] Rich message text and image buttons (#11473)

* Added text and image buttons

Added image button with msg

Fixed according to updated JSON

Added horizontal buttons support

Fixed indentation

Added text and image buttons

* Added image button with msg

* Fixed according to updated JSON

* Added horizontal buttons support

* fix style logic

* Fixed lint errors and merge conflict

* Remove lint errors

* Added Actions schema to tests

renatobecker committed:
[FIX] Livechat open room method (#11830)

renatobecker committed:
[FIX] Livechat open room method (#11830)

ura14h committed:
[FIX] Fix links in `onTableItemClick` of the directroy page (#11543)

* Fix links in `onTableItemClick` of the directroy page

* Fix an incorrect indent

ura14h committed:
[FIX] Fix permalink of message when running system with subdir (#11781)

* Fix permalink of message when running system with subdir

* Fix linter error

* fix some code style

Commit:
[FIX] Searching by `undefined` via REST when using `query` param (#11657)

ggazzo committed:
permissions name no break (#11836)

Atisom committed:
[FIX] minor fixes in hungarian i18n (#11797)

[FIX] minor fixes in hungarian i18n

rssilva committed:
[FIX] re-adding margin to menu icon on header (#11778)

engelgabriel committed:
LingoHub Update :rocket: (#11838)

Manual push by LingoHub User: Rodrigo Nascimento.
Project: Rocket.Chat

Made with :heart: by https://lingohub.com

Commit:
[FIX] Results pagination on /directory REST endpoint (#11551)

renatobecker committed:
[FIX] Livechat rooms starting with two unread message counter (#11834)

jukper committed:
Fixed the finnish translation and remove some profanities. (#11794)

TheReal1604 committed:
Fixed message pruning translations (#11691)

Hudell committed:
[NEW] Setting to enable/disable slack bridge reactions (#10217)

Commit:
[NEW] REST endpoint to manage server assets (#11697)

Commit:
[FIX] REST `im.members` endpoint not working without sort parameter (#11821)

Hudell committed:
[FIX] Revoked `view-d-room` permission logics (#11522)

Commit:
[NEW] Personal access tokens for users to create API tokens (#11638)

renatobecker committed:
[IMPROVE] Start storing Livechat department within rooms (#11733)

rodrigok committed:
[FIX] Some assets were pointing to nonexistent path (#11796)

Hudell committed:
[NEW] Slackbridge: send attachment notifications (#10269)

renatobecker committed:
[FIX] LiveChat switch department not working (#11011)

rodrigok committed:
[NEW] Beta support for Big Blue Button video conferencing system (#11837)

Hudell committed:
[FIX] Delete removed user's subscriptions (#10700)

* Remove subscriptions when a user is deleted and add migration to delete past subscriptions

* abort user deletion when they are the last owner of a room

sampaiodiego committed:
Bump version to 0.69.0-rc.0

pierre-lehnen-rc committed:
reduced code complexity

pierre-lehnen-rc committed:
Reduced code complexity

pierre-lehnen-rc committed:
Added OAuth fixes for WeChat

pierre-lehnen-rc committed:
eslint

pierre-lehnen-rc committed:
moved assertions up

sampaiodiego committed:
Bump version to 0.68.5

engelgabriel commented:
Hi @lhr0909 this looks very cool! Can you please open a Pull Request to our main project so we can collaborate and merge? it?

sampaiodiego committed:
Merge pull request #11852 from RocketChat/release-0.68.5

Release 0.68.5

c19 committed:
set email to <openid>@wechat.app

pierre-lehnen-rc committed:
Added WeChat OAuth class

Hudell commented:
I went ahead and opened a [WIP] PR including the commits from this PR and some new code I wrote:
https://github.com/RocketChat/Rocket.Chat/pull/11856

I'm creating a new class for the WeChat OAuth, so that it can change the behaviour of the login process to include the new popup and the other non-default stuff.

pierre-lehnen-rc committed:
Added WeChat icon

pierre-lehnen-rc committed:
Working on the WeChat OAuth class

pierre-lehnen-rc committed:
Removed unused Strings

pierre-lehnen-rc committed:
Configure wechat email

pierre-lehnen-rc committed:
eslint

ggazzo committed:
Regression: role tag background, unread item font and message box autogrow (#11861)

sampaiodiego committed:
Fix final page wizard logo

rodrigok committed:
[FIX] Translations were not unique per app allowing conflicts among apps (#11878)

* [FIX] Translations were not unique per app allowing conflicts among apps

* Remove console.log

* Load correct translations when a new app is added

rodrigok committed:
[NEW] Internal marketplace for apps (#11864)

* Update page details

* Adds icon to marketplace install button

* Add icon and animation for installing apps

* Fixes animation svg for app installation

* Changes to actions

* Fixing conflicts

* Adding the right return to apps()

* Fetching apps according tab context ('marketplace' or 'installed')

* Set loading on tables

* Adding POST to install the app on click

* Fixing linting

* Tagging already installed apps with prop '_installed' to show the 'v' icon

* Adding helper to render the 'v' icon when the app is already downloaded

* Commenting _appOnAppAdded method for now

* Fixing arrow functions returning objects

* Adding isInstalled helper to appManage to hide 'install' button on details page

* Linting

* Fixing event propagation on table, adding uninstall button

* Removing unnecessary comments

* Fix missing image for isntalled apps

* Add translation for Apps_Details

* Improvements on app details screen

* Show categories for installed apps

* Allow edit app settings

* Improve app settings layout

* Allow update apps

* Small layout fixes for marketplace

* some css fixes

* Adding marketplaceApiVersion to rockethat.info

* Replacing hardcoded colors for vars and removing \!important

* Removing unused class, removing \!important

* Fixing page scroll

* codacy

rodrigok committed:
App engine merge (#11835)

rodrigok committed:
Update marketplaceApiVersion to 1.0.0

rodrigok committed:
Fix apps install from apps list

rodrigok committed:
Fix some layout of the internal marketplace

sampaiodiego committed:
Merge remote-tracking branch 'origin/develop' into release-candidate

sampaiodiego committed:
Bump version to 0.69.0-rc.1

sampaiodiego committed:
Merge remote-tracking branch 'origin/release-candidate' into release-0.69.0

sampaiodiego committed:
Bump version to 0.69.0

sampaiodiego committed:
Merge pull request #11881 from RocketChat/release-0.69.0

Release 0.69.0

rodrigok committed:
[FIX] Hipchat import was failing when importing messages from a non existent user (#11892)

rodrigok committed:
[FIX] Hipchat import was failing when importing messages from a non existent user (#11892)

rodrigok committed:
[FIX] Hipchat importer was not importing users without emails and uploaded files (#11910)

rodrigok committed:
[FIX] Hipchat importer was not importing users without emails and uploaded files (#11910)

rodrigok committed:
[FIX] App updates were not being shown correctly (#11893)

rodrigok committed:
[FIX] App updates were not being shown correctly (#11893)

ubarsaiyan committed:
[FIX] Duplicated message buttons (#11853)

ubarsaiyan committed:
[FIX] Duplicated message buttons (#11853)

sampaiodiego committed:
Bump version to 0.69.1

sampaiodiego committed:
Merge pull request #11918 from RocketChat/release-0.69.1

Release 0.69.1

sampaiodiego committed:
Merge remote-tracking branch 'origin/master' into develop-sync

sampaiodiego committed:
Bump version to 0.70.0-develop

sampaiodiego committed:
Merge pull request #11921 from RocketChat/develop-sync

Merge master into develop & Set version to 0.70.0-develop

engelgabriel committed:
adding integrations-attachments-example.png

TobiasKappe committed:
[FIX] Typo in a configuration key for SlackBridge excluded bot names (#11872)

renatobecker committed:
[FIX] Missing `fname` property in Livechat rooms (#11849_

thaiphv committed:
[BREAK] Update the default port of the Prometheus exporter (#11351)

MartinSchoeler committed:
[FIX] Close popover on shortcuts and writing (#11562)

der-eismann committed:
[IMPROVE] Refactoring the snap backup scripts to improve UX

lhr0909 committed:
try to keep hard-coded wechat settings away from users

mrsimpson committed:
[NEW] Informal German translations (#9984)

* Provide a dedicated informal german language option

* Use a "virtual" language code de-IN for informal german
The IETF language specification compliant suffix "informal"  did not work with the i18n library used

timkinnane committed:
[NEW] Include room name in stream for bots (#11812)

timkinnane committed:
[FIX] Permission check on joinRoom for private room (#11857)

geekgonecrazy committed:
[FIX] Wrong build path in install.sh (#11879)

rodrigok committed:
[FIX] Reset password link error if already logged in (#12022)

cardoso committed:
[FIX] Apps: setting with 'code' type only saving last line (#11992)

kaiiiiiiiii committed:
[FIX] Update user information not possible by admin if disabled to users (#11955)

ggazzo committed:
[FIX] Hidden admin sidenav on embedded layout (#12025)

aferreira44 committed:
[NEW] REST endpoints to create roles and assign roles to users (#11855)

Hudell committed:
Merge branch 'develop' into wechat-oauth
